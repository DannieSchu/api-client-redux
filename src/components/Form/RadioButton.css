.RadioButton {  
  input[type="radio"] {
    display: none;
  }

  input[type="radio"]:checked + label {
    background-color: #ee26cd;
}

  label:hover {
    background-color: #ec66d6;
  }

  label {
    width: 8rem;
    min-width: 4.8rem;
    height: 2.8em;
    border-radius: .3rem;

    font-size: 1.2rem;
    line-height: 2.8em;
    vertical-align: middle;
    text-align: center;
    text-transform: uppercase;
    
    color: white;
    background-color: rgb(180, 180, 180);
    cursor: pointer;
  }
}
