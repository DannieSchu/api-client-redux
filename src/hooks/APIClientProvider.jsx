import React, { createContext, useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { apiData } from '../services/apiData.jsx';

const APIClientContext = createContext();

export const APIClientProvider = ({ children }) => {
  const [method, setMethod] = useState('get');
  const [url, setUrl] = useState('');
  const [body, setBody] = useState('');
  const [results, setResults] = useState(null);
  const [requests, setRequests] = useState([]);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    const storage = JSON.parse(localStorage.getItem('requests'));
    if(storage) setRequests(storage);
  }, []);

  const stateFactory = {
    url: setUrl,
    method: setMethod,
    body: setBody
  };

  const handleChange = ({ target }) => stateFactory[target.name](target.value);

  const handleSubmit = e => {
    e.preventDefault();
    setLoading(true);
    apiData(url, method, body)
      .then(res => {
        setResults(res),
        setLoading(false);
      });
    setRequests(prevRequests => {
      if(prevRequests.filter(request => request.url === url && request.method === method).length < 1) {
        return [{
          url,
          method,
          body: body || null
        }, ...prevRequests];
      } else {
        return prevRequests;
      }
    });
    localStorage.setItem('requests', JSON.stringify(requests));
  };

  const handleClick = ({ url, method, body }) => {
    setUrl(url);
    setMethod(method);
    setBody(body || null);
    setResults(null);
  };

  const handleDelete = () => {
    localStorage.clear();
    setRequests([]);
  };

  return (
    <APIClientContext.Provider value={{
      url,
      method,
      body,
      results,
      requests,
      loading,
      handleChange,
      handleSubmit,
      handleClick,
      handleDelete
    }}>
      {children}
    </APIClientContext.Provider>
  );
};

APIClientProvider.propTypes = {
  children: PropTypes.node
};

export const useRequestContext = () => {
  const context = useContext(APIClientContext);
  return context;
};

export const useUrl = () => {
  const { url } = useRequestContext();
  return url;
};

export const useMethod = () => {
  const { method } = useRequestContext();
  return method;
};

export const useBody = () => {
  const { body } = useRequestContext();
  return body;
};

export const useLoading = () => {
  const { loading } = useRequestContext();
  return loading;
};

export const useResults = () => {
  const { results } = useRequestContext();
  return results;
};

export const useRequests = () => {
  const { requests } = useRequestContext();
  return requests;
};

export const useHandleChange = () => {
  const { handleChange } = useRequestContext();
  return handleChange;
};

export const useHandleSubmit = () => {
  const { handleSubmit } = useRequestContext();
  return handleSubmit;
};

export const useHandleClick = () => {
  const { handleClick } = useRequestContext();
  return handleClick;
};

export const useHandleDelete = () => {
  const { handleDelete } = useRequestContext();
  return handleDelete;
};
