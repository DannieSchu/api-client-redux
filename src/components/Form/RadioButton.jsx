import React from 'react';
import PropTypes from 'prop-types';
import styles from './RadioButton.css';
import { useHandleChange } from '../../hooks/APIClientProvider';

export const RadioButton = ({ value }) => {
  const handleChange = useHandleChange();

  return (
    <div className={styles.RadioButton}>
      <input
        id={value}
        type="radio"
        name="method"
        value={value}
        onChange={handleChange}
      />
      <label htmlFor={value}>{value}</label>
    </div>
  );
};

RadioButton.propTypes = {
  value: PropTypes.string.isRequired
};
