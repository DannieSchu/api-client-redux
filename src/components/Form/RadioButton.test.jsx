import React from 'react';
import { shallow } from 'enzyme';
import { RadioButton } from './RadioButton.jsx';
import { APIClientProvider } from '../../hooks/APIClientProvider.jsx';

describe('RadioButton', () => {
  it('matches a snapshot', () => {
    const wrapper = shallow(
      <APIClientProvider>
        <RadioButton
          value="get" />
      </APIClientProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
