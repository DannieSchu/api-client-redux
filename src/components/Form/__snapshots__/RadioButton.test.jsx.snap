// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton matches a snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "body": "",
      "handleChange": [Function],
      "handleClick": [Function],
      "handleDelete": [Function],
      "handleSubmit": [Function],
      "loading": false,
      "method": "get",
      "requests": Array [],
      "results": null,
      "url": "",
    }
  }
>
  <RadioButton
    value="get"
  />
</ContextProvider>
`;
